src/pretix/base/timeframes.py

```diff
@@ -0,0 +1,388 @@
+
+    def decompress(self, value):
+        if not value:
+            return ['unset', None, None]
+        if '/' in value:
+            return [
+                'custom',
+                date.fromisoformat(value.split('/', 1)[0]),
+                date.fromisoformat(value.split('/', 1)[-1]),
+            ]
+        return []
+
+    def get_context(self, name, value, attrs):
+        ctx = super().get_context(name, value, attrs)
+        ctx['required'] = self.timeframe_choices[0][0] == 'unset'
+        return ctx
+
+
+def _describe_timeframe(label, start, end, future, describe):
+    details = f'{date_format(start(now()), "SHORT_DATE_FORMAT")} - {date_format(end(now(), start(now())), "SHORT_DATE_FORMAT")}'
+    if describe:
+        details = f'{describe(start(now()))}: {details}'
+    return f'{label} ({details})'
+
+
+class DateFrameField(forms.MultiValueField):
+    def __init__(self, *args, **kwargs):
+        include_future_frames = kwargs.pop('include_future_frames')
+
+        top_choices = [('custom', gettext_lazy('Custom timeframe'))]
+        if not kwargs.get('required', True):
+            top_choices.insert(0, ('unset', pgettext_lazy('reporting_timeframe', 'All time')))
+
+        _choices = []
+        for grouper, group in groupby(REPORTING_DATE_TIMEFRAMES, key=lambda i: i[5]):
+            _choices.append((grouper, [
+                (identifier, _describe_timeframe(label, start, end, future, describe))
+                for identifier, label, start, end, future, group, describe in group
+                if include_future_frames or not future
+            ]))
+
+        timeframe_choices = [
+            ('', top_choices)
+        ] + _choices
+
+        fields = (
+            forms.ChoiceField(
+                choices=timeframe_choices,
+                required=True
+            ),
+            forms.DateField(
+                required=False
+            ),
+            forms.DateField(
+                required=False
+            ),
+        )
+        if 'widget' not in kwargs:
+            kwargs['widget'] = DateFrameWidget(timeframe_choices=timeframe_choices)
+        kwargs.pop('max_length', 0)
+        kwargs.pop('empty_value', 0)
+        super().__init__(
+            fields=fields, require_all_fields=False, *args, **kwargs
+        )
+
+    def compress(self, data_list):
+        if not data_list:
+            return None
+        if data_list[0] == 'unset':
+            return None
+        elif data_list[0] == 'custom':
+            return f'{data_list[1].isoformat()}/{data_list[2].isoformat()}'
+        else:
+            return data_list[0]
+
+    def has_changed(self, initial, data):
+        if initial is None:
+            initial = self.widget.decompress(initial)
+        return super().has_changed(initial, data)
+
+    def clean(self, value):
+        if value[0] == 'custom':
+            if not value[1] or not value[2]:
+                raise ValidationError(self.error_messages['incomplete'])
```

Comment:
Would it make sense to allow incomplete date-ranges? That way one could generate reports for everything up until a certain date or only beginning on a certain date. I would not auto-create a valid start-date or end-date if none is given, but patch time-range parsing. See my other comment.